<?xml version="1.0" ?>
<robot name="robot" xmlns:xacro="http://ros.org/wiki/xacro">

  <!-- ****************** MATH CONSTANTS *******************************  -->
  <xacro:property name="PI_2" value="1.57079632679"/>
  <xacro:property name="PI" value="3.141592654"/>
  <!-- ****************** ROBOT CONSTANTS *******************************  -->
  <!-- Define the size of the robot's main chassis in meters -->
  <xacro:property name="base_radius" value="0.05"/>
  <xacro:property name="base_height" value="0.05"/>
  <xacro:property name="link_width" value="0.1"/>
  <xacro:property name="link_height" value="0.07"/>
  <xacro:property name="link_1_length" value="0.02"/>
  <xacro:property name="link_2_length" value="0.0825"/>
  <xacro:property name="link_3_length" value="0.086"/>
  <xacro:property name="link_4_length" value="0.075"/>
	
  <link name="world"/>

  <!-- ********************** ROBOT BASE *********************************  -->
  <link name="base">
    <visual>
      <origin xyz="0 0 0" rpy="${PI_2} 0.0 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/Base.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 ${0.5*base_height}"/>
      <geometry>
        <cylinder radius="${base_radius}" length="${base_height}"/>
      </geometry>
    </collision>
  </link>
  
  <!-- ********************** LINK 1 *********************************  -->
  <link name="link1">
    <visual>
      <origin xyz="0 0 -0.05" rpy="${PI_2} 0.0 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/Link1.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 ${link_1_length/2}" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="${link_width} ${link_height} ${link_1_length}"/>
      </geometry>
    </collision>
  </link>

  <!-- ********************** LINK 2 *********************************  -->
  <link name="link2">
    <visual>
      <origin xyz="0 0 -0.07" rpy="${PI_2} 0.0 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/Link2.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 ${link_2_length/2}" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="${link_width} ${link_height} ${link_2_length}"/>
      </geometry>
    </collision>
  </link>

  <!-- ********************** LINK 3 *********************************  -->
  <link name="link3">
    <visual>
      <origin xyz="0 0.005 -0.1525" rpy="${PI_2} 0.0 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/Link3.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 ${link_3_length/2}" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="${link_width} ${link_height} ${link_3_length}"/>
      </geometry>
    </collision>
  </link>

  <!-- ********************** LINK 4 *********************************  -->
  <link name="link4">
    <visual>
      <origin xyz="0 -0.006 -0.2375" rpy="${PI_2} 0.0 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/Link4.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 0 ${link_4_length/2}" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="${link_width} ${link_height} ${link_4_length}"/>
      </geometry>
    </collision>
  </link>

  <!-- ********************** Gripper *********************************  -->
  <link name="gripper_left">
    <visual>
      <origin xyz="0 0 0" rpy="${-PI_2} ${-PI_2} 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/GripperLeft.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 ${-link_width/4} ${link_4_length/4}" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="${link_width/2} ${link_height/4} ${link_4_length/2}"/>
      </geometry>
    </collision>
  </link>

  <link name="gripper_right">
    <visual>
      <origin xyz="0 -0.005 0" rpy="${-PI_2} ${-PI_2} 0.0"/>
      <geometry>
        <mesh filename="package://edubot/meshes/GripperRight.dae" scale="1 1 1"/>
      </geometry>
    </visual>

    <collision>
      <origin xyz="0 ${link_width/4} ${link_4_length/4}" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="${link_width/2} ${link_height/4} ${link_4_length/2}"/>
      </geometry>
    </collision>
  </link>

  <!-- **************** ROBOT EE *************************************** -->
  <link name="ee"/>

  <!-- ********************** Joints *********************************  -->
  <joint name="world_joint" type="fixed">
    <parent link="world"/>
    <child link="base" />
    <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
  </joint>

  <joint name="link1_joint" type="revolute">
    <parent link="base"/>
    <child link="link1"/>
    <axis xyz="0 0 1"/>
    <limit lower="${-PI_2}" upper="${PI_2}" effort="10" velocity="10"/>
    <origin xyz="0 0 ${base_height}" rpy="0 0 0"/>
  </joint>

  <joint name="link2_joint" type="revolute">
    <parent link="link1"/>
    <child link="link2"/>
    <axis xyz="0 1 0"/>
    <limit lower="${-PI_2}" upper="${PI_2}" effort="10" velocity="10"/>
    <origin xyz="0 0 ${link_1_length}" rpy="0 0 0"/>
  </joint>

  <joint name="link3_joint" type="revolute">
    <parent link="link2"/>
    <child link="link3"/>
    <axis xyz="0 1 0"/>
    <limit lower="${-PI_2}" upper="${PI_2}" effort="10" velocity="10"/>
    <origin xyz="0 0 ${link_2_length}" rpy="0 ${PI_2} ${PI}"/>
  </joint>

  <joint name="link4_joint" type="revolute">
    <parent link="link3"/>
    <child link="link4"/>
    <axis xyz="0 1 0"/>
    <limit lower="${-PI_2}" upper="${PI_2}" effort="10" velocity="10"/>
    <origin xyz="0 0 ${link_3_length}" rpy="0 0 ${-PI}"/>
  </joint>

  <joint name="ee_joint" type="fixed">
    <parent link="link4"/>
    <child link="ee"/>\
    <origin xyz="0 0 ${link_4_length}" rpy="0 0 0"/>
  </joint>

  <joint name="gripper_left_joint" type="prismatic">
    <parent link="ee"/>
    <child link="gripper_left"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0 -0.01 -0.025" rpy="0 0 0"/>
    <limit lower="${-0.0}" upper="${0.02}" effort="10" velocity="10"/>
  </joint> 

  <joint name="gripper_right_joint" type="prismatic">
    <parent link="ee"/>
    <child link="gripper_right"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0 0.01 -0.025" rpy="0 0 0"/>
    <limit lower="${-0.0}" upper="${0.02}" effort="10" velocity="10"/>
    <mimic joint="gripper_left_joint" multiplier="-1" offset="0" />
  </joint>  

</robot>